### PR TITLE
Fixes for running on Windows

### PR DESCRIPTION
**Pull request by Houdinic**
ebookmaker windows fix

### CODE REVIEW

**Comment by eshellman** [comment]
maybe @windymilla can take a look

File: ebookmaker/EbookMaker.py [hunk 1]
```diff
@@ -273,149 +276,153 @@ def add_local_options (ap):
...
+        log_handler = open_log(os.path.join(job.outputdir, job.logfile))
 
-    debug ('=== Building %s ===' % job.type)
+    debug('=== Building %s ===' % job.type)
 
     try:
         if job.url:
-            spider = Spider.Spider ()
+            spider = Spider.Spider()
+            if job.url[:4]=='http':
```
Comment (eshellman) [inline ebookmaker/EbookMaker.py:416]: any reason not to use job.url.startswith('http') ?

File: ebookmaker/EbookMaker.py [hunk 2]
```diff
@@ -273,149 +276,153 @@ def add_local_options (ap):
...
+                sep = '/'
+            else:
+                sep=os.path.sep
```
Comment (eshellman) [inline ebookmaker/EbookMaker.py:419]: PEP8:
sep = os.path.sep

File: ebookmaker/EbookMaker.py (hunk 1 above)
Comment (Houdinic) [reply ebookmaker/EbookMaker.py:416]: I thought startwith() is sudo code.

**Comment by windymilla** [comment]
I don't really have the knowledge to comment on the edits, but I'm happy to try a new version of ebookmaker when it's at a point I can install it. Would I just need to redo "pipenv install ebookmaker"?

**Comment by eshellman** [comment]
Yes please!
Should be
`pipenv uninstall ebookmaker
pipenv install git+https://github.com/gutenbergtools/ebookmaker.git@winpath#egg=ebookmaker`

File: ebookmaker/EbookMaker.py (hunk 1 above)
Comment (eshellman) [reply ebookmaker/EbookMaker.py:416]: Haha that's why I like Python.

**Comment by windymilla** [comment]
I did pipenv shell, then pipenv uninstall ebookmaker
The uninstall worked fine.

I then did
pipenv install git+
https://github.com/gutenbergtools/ebookmaker.git@winpath#egg=ebookmaker
with the following results (apologies for the formatting - that's how it
appeared in the command window)

Installing git+
https://github.com/gutenbergtools/ebookmaker.git@winpath#egg=ebookmaker.
Warning: You installed a VCS dependency in non-editable mode. This will
work fine, but sub-dependencies will not be reso
lved by $ pipenv lock.
  To enable this sub-dependency functionality, specify that this dependency
is editable.
Collecting ebookmaker
  Cloning https://github.com/gutenbergtools/ebookmaker.git (to revision
winpath) to c:\users\nblower\appdata\local\temp\
pip-install-0_68g52e\ebookmaker

Error:  An error occurred while installing git+
https://github.com/gutenbergtools/ebookmaker.git@winpath#egg=ebookmaker!
b'  Running command git clone -q
https://github.com/gutenbergtools/ebookmaker.git
\'C:\\Users\\nblower\\AppData\\Local\\
Temp\\pip-install-0_68g52e\\ebookmaker\'\n  Running command git checkout -b
winpath --track origin/winpath\n  Branch \'w
inpath\' set up to track remote branch \'winpath\' from \'origin\'.\n
 Switched to a new branch \'winpath\'\n    ERROR:
Command errored out with exit status 1:\n     command:
\'c:\\users\\nblower\\.virtualenvs\\ebookmaker-cgaquyhi\\scripts\
\python.exe\' -c \'import sys, setuptools, tokenize; sys.argv[0] =
\'"\'"\'C:\\\\Users\\\\nblower\\\\AppData\\\\Local\\\
\Temp\\\\pip-install-0_68g52e\\\\ebookmaker\\\\setup.py\'"\'"\';
__file__=\'"\'"\'C:\\\\Users\\\\nblower\\\\AppData\\\\L
ocal\\\\Temp\\\\pip-install-0_68g52e\\\\ebookmaker\\\\setup.py\'"\'"\';f=getattr(tokenize,
\'"\'"\'open\'"\'"\', open)(_
_file__);code=f.read().replace(\'"\'"\'\\r\\n\'"\'"\',
\'"\'"\'\\n\'"\'"\');f.close();exec(compile(code, __file__, \'"\'
"\'exec\'"\'"\'))\' egg_info --egg-base
\'C:\\Users\\nblower\\AppData\\Local\\Temp\\pip-install-0_68g52e\\ebookmaker\\pi
p-egg-info\'\n         cwd:
C:\\Users\\nblower\\AppData\\Local\\Temp\\pip-install-0_68g52e\\ebookmaker\\\n
   Complete o
utput (7 lines):\n    Traceback (most recent call last):\n      File
"<string>", line 1, in <module>\n      File "C:\\Us
ers\\nblower\\AppData\\Local\\Temp\\pip-install-0_68g52e\\ebookmaker\\setup.py",
line 67, in <module>\n        long_desc
ription = open (\'README.md\').read (),\n      File
"c:\\users\\nblower\\.virtualenvs\\ebookmaker-cgaquyhi\\lib\\encodin
gs\\cp1252.py", line 23, in decode\n        return
codecs.charmap_decode(input,self.errors,decoding_table)[0]\n    Unico
deDecodeError: \'charmap\' codec can\'t decode byte 0x9d in position 4066:
character maps to <undefined>\n    ----------
------------------------------\nERROR: Command errored out with exit status
1: python setup.py egg_info Check the logs f
or full command output.\n'
This is likely caused by a bug in ebookmaker. Report this to its
maintainers.
Installation Failed

**Comment by windymilla** [comment]
I did pipenv shell, then pipenv uninstall ebookmaker
The uninstall worked fine.

I then did
pipenv install git+https://github.com/gutenbergtools/ebookmaker.git@winpath#egg=ebookmaker
with the following results (apologies for formatting - copy/pasted error from command window)

`Installing git+https://github.com/gutenbergtools/ebookmaker.git@winpath#egg=ebookmaker.
Warning: You installed a VCS dependency in non-editable mode. This will work fine, but sub-dependencies will not be reso
lved by $ pipenv lock.
  To enable this sub-dependency functionality, specify that this dependency is editable.
Collecting ebookmaker
  Cloning https://github.com/gutenbergtools/ebookmaker.git (to revision winpath) to c:\users\nblower\appdata\local\temp\
pip-install-0_68g52e\ebookmaker

Error:  An error occurred while installing git+https://github.com/gutenbergtools/ebookmaker.git@winpath#egg=ebookmaker!
b'  Running command git clone -q https://github.com/gutenbergtools/ebookmaker.git \'C:\\Users\\nblower\\AppData\\Local\\
Temp\\pip-install-0_68g52e\\ebookmaker\'\n  Running command git checkout -b winpath --track origin/winpath\n  Branch \'w
inpath\' set up to track remote branch \'winpath\' from \'origin\'.\n  Switched to a new branch \'winpath\'\n    ERROR:
Command errored out with exit status 1:\n     command: \'c:\\users\\nblower\\.virtualenvs\\ebookmaker-cgaquyhi\\scripts\
\python.exe\' -c \'import sys, setuptools, tokenize; sys.argv[0] = \'"\'"\'C:\\\\Users\\\\nblower\\\\AppData\\\\Local\\\
\Temp\\\\pip-install-0_68g52e\\\\ebookmaker\\\\setup.py\'"\'"\'; __file__=\'"\'"\'C:\\\\Users\\\\nblower\\\\AppData\\\\L
ocal\\\\Temp\\\\pip-install-0_68g52e\\\\ebookmaker\\\\setup.py\'"\'"\';f=getattr(tokenize, \'"\'"\'open\'"\'"\', open)(_
_file__);code=f.read().replace(\'"\'"\'\\r\\n\'"\'"\', \'"\'"\'\\n\'"\'"\');f.close();exec(compile(code, __file__, \'"\'
"\'exec\'"\'"\'))\' egg_info --egg-base \'C:\\Users\\nblower\\AppData\\Local\\Temp\\pip-install-0_68g52e\\ebookmaker\\pi
p-egg-info\'\n         cwd: C:\\Users\\nblower\\AppData\\Local\\Temp\\pip-install-0_68g52e\\ebookmaker\\\n    Complete o
utput (7 lines):\n    Traceback (most recent call last):\n      File "<string>", line 1, in <module>\n      File "C:\\Us
ers\\nblower\\AppData\\Local\\Temp\\pip-install-0_68g52e\\ebookmaker\\setup.py", line 67, in <module>\n        long_desc
ription = open (\'README.md\').read (),\n      File "c:\\users\\nblower\\.virtualenvs\\ebookmaker-cgaquyhi\\lib\\encodin
gs\\cp1252.py", line 23, in decode\n        return codecs.charmap_decode(input,self.errors,decoding_table)[0]\n    Unico
deDecodeError: \'charmap\' codec can\'t decode byte 0x9d in position 4066: character maps to <undefined>\n    ----------
------------------------------\nERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs f
or full command output.\n'
This is likely caused by a bug in ebookmaker. Report this to its maintainers.
Installation Failed`

**Comment by eshellman** [comment]
ok, that's interesting. You can silence the first warning with 
`pipenv install -e git+https://github.com/gutenbergtools/ebookmaker.git@winpath#egg=ebookmaker`

We'll see if we can reproduce and fix the second error - looks like it could be an issue with platform-dependent default encoding.

**Comment by eshellman** [comment]
@windymilla @Houdinic could you run 
`python -c "import locale;print(locale.getpreferredencoding())"`
and report the results

**Comment by eshellman** [comment]
@windymilla  Thanks so much for helping us test this. We believe we've fixed all the issues running  ebookmaker on Windows, but can only test on Windows 10.

**Comment by windymilla** [comment]
That's getting closer for me thanks. Looks like I have a problem with tidy. Do I have to install that separately for ebookmaker to use? This is all new to me.

(ebookmaker-cgaQuYhi) C:\DP\ebookmaker>pipenv run python C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\Scripts\ebook
maker -v -v --make=epub.images --output-dir=/DP/ebookmaker ./nature_study.htm
DEBUG:root:Building types: epub.images
DEBUG:root:Loading parser from module: AuxParser for mediatypes: */*
DEBUG:root:Loading parser from module: CSSParser for mediatypes: text/css
WARNING:root:No gnu dbm support found. Rhyming dictionary not used.
DEBUG:root:Loading parser from module: GutenbergTextParser for mediatypes: text/plain
DEBUG:root:Loading parser from module: HTMLParser for mediatypes: text/html, application/xhtml+xml
DEBUG:root:Loading parser from module: ImageParser for mediatypes: image/jpeg, image/png, image/gif
DEBUG:root:Loading parser from module: RSTParser for mediatypes: text/x-rst
DEBUG:root:Loading writer type epub from module EpubWriter
DEBUG:root:Loading writer type html from module HTMLWriter
DEBUG:root:Loading writer type kindle from module KindleWriter
DEBUG:root:Loading writer type pdf from module PDFWriter
DEBUG:root:Loading writer type picsdir from module PicsDirWriter
DEBUG:root:Loading writer type rst from module RSTWriter
DEBUG:root:Loading writer type txt from module TxtWriter
DEBUG:root:Loading packager type: gzip from module: GzipPackager for formats: rst, html.noimages, html.images, txt.us-as
cii, txt.iso-8859-1, txt.utf-8
DEBUG:root:Loading packager type: ww from module: HTMLPackager for formats: html.images
DEBUG:root:Loading packager type: ww from module: PDFPackager for formats:
DEBUG:root:Loading packager type: ww from module: PushPackager for formats: push
DEBUG:root:Loading packager type: ww from module: RSTPackager for formats: rst.gen
DEBUG:root:Loading packager type: ww from module: TxtPackager for formats: txt.us-ascii, txt.iso-8859-1, txt.utf-8
DEBUG:root:... got mediatype text/html from guess_type
DEBUG:root:... creating new parser for C:\DP\ebookmaker\nature_study.htm
DEBUG:root:HTMLParser.parse () ...
DEBUG:root:HTMLParser.pre_parse () ...
DEBUG:root:Fetching C:\DP\ebookmaker\nature_study.htm ...
DEBUG:root:Got charset utf-8 from html meta
DEBUG:root:Trying to decode document with charset utf_8_sig ...
INFO:root:Running html thru tidy.
Traceback (most recent call last):
  File "C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\Scripts\ebookmaker", line 18, in <module>
    EbookMaker.main ()
  File "C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\lib\site-packages\ebookmaker\EbookMaker.py", line 524, in main

    options.dc = get_dc(options.url) # this is when doc at url gets parsed!
  File "C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\lib\site-packages\ebookmaker\EbookMaker.py", line 170, in get_
dc
    parser.parse()
  File "C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\lib\site-packages\ebookmaker\parsers\HTMLParser.py", line 455,
 in parse
    self.pre_parse ()
  File "C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\lib\site-packages\ebookmaker\parsers\HTMLParser.py", line 437,
 in pre_parse
    html = self.tidy (html)
  File "C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\lib\site-packages\ebookmaker\parsers\HTMLParser.py", line 157,
 in tidy
    stderr = subprocess.PIPE)
  File "C:\Python\Lib\subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "C:\Python\Lib\subprocess.py", line 1178, in _execute_child
    startupinfo)
FileNotFoundError: [WinError 2] The system cannot find the file specified

**Comment by eshellman** [comment]
yes, you need to install tidy

**Comment by windymilla** [comment]
Thanks. I had tidy installed for use in Guiguts, so just added its path to the PATH.

I can now create an ePub file, which I was able to open in Calibre!
However, I can't open it in ADE ("Errors were encountered in <filename>")
I used online ebookmaker to create an epub from the same HTML & images and that opens fine in ADE.
Here are the output log from my local run of ebookmaker, the epub file created locally and the epub file created online. When I unzip the two epub files I can see the structure is quite different - is that expected, and could it be the cause of the error in ADE?
[output.zip](https://github.com/gutenbergtools/ebookmaker/files/4471120/output.zip)

**Comment by windymilla** [comment]
Success with epub - file opens fine in ADE now! 
Many thanks for working on this @Houdinic @eshellman 

Next, I tried creating the Kindle file too. First attempt failed because I didn't have path to kindlegen executable on my PATH, so I fixed that. 
The end of my output log looks like this now:

> DEBUG:root:=== Building kindle.images ===
> DEBUG:root:Start of retrieval
> DEBUG:root:Dropping not included mediatype application/epub+zip
> DEBUG:root:End of retrieval
> INFO:root:Creating Kindle file: /DP/ebookmaker\UnknownTitle-images-kindle.mobi
> INFO:root:            ... from: /DP/ebookmaker\UnknownTitle-images-epub.epub
> INFO:root:Done Kindle file: /DP/ebookmaker\UnknownTitle-images-kindle.mobi

However, the mobi file doesn't appear in the folder.

I typed into my command window:
kindlegen UnknownTitle-images-epub.epub
and it created the mobi file

So kindlegen can create the file, but isn't doing it when spawned from within ebookmaker.
Any ideas?
Thanks, Nigel

**Comment by eshellman** [comment]
probably the mobi file is in a folder named "/DP/ebookmaker". also you might try running with option --output-dir=C:\DP\ebookmaker to set a windows style path explicitly

**Comment by windymilla** [comment]
Thanks - it didn't work with your latest commit, but it did work when I set the output-dir to a windows style path explicitly. This was my command:
pipenv run python C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\Scripts\ebookmaker -v -v --make=epub.images --make=kindle.images --output-dir=\DP\ebookmaker .\test.html

I think that can now be said to be working - well done guys!

I created a Windows .bat file, containing

> cd C:\DP\ebookmaker
> pipenv run python C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\Scripts\ebookmaker -v --make=epub.images --make=kindle.images --output-dir=%1 --title=%2 %3
 
then in Guiguts, set up a Custom External Operation with the command

> c:\dp\ebookmaker\run_ebookmaker.bat $d $f $d$f$e

Then I could run ebookmaker from inside Guiguts, which was a previous feature that was requested that triggered me looking into installing ebookmaker under Windows.

One thing I noticed was that things go wrong with ebookmaker (& maybe kindlegen) if your folder path has spaces in its name, even with quotes around the name. That however is a limitation that I can live with.

**Comment by windymilla** [comment]
I think it might be possible to simplify the install/running notes for Windows.
I don't think No. 9, with pipenv --three is necessary, nor No. 10, pipenv shell (that doesn't seem to give any advantage as far as I can see)

Also, we need to say that the paths to the tidy and kindlegen executables both need to be added to system PATH env variable.

Also, is it possible for ebookmaker to run without Cairo? Do I understand correctly that it is only used for generating a missing cover? A user running locally wouldn't really need that, and a DP user would always have a cover. It would be nice if the user didn't need to install Cairo and ebookmaker just used a default cover if Cairo wasn't there. No idea how much work that would be.  It would then also mean 12, 13, 14, 15 & 17 could go from the instructions.

Finally, we should add an example command for how to actually generate an epub file, maybe based on the contents of the .bat file mentioned in the previous comment.

**Comment by eshellman** [comment]
It should run without Cairo now, turns cover generation off - we added some exception handling. Not sure we've properly tested.

**Comment by windymilla** [comment]
I renamed the folder I had installed gtk in (for Cairo).
When I ran ebookmaker it crashed with a traceback.
I uninstalled ebookmaker and installed the latest version using pipenv
However when I ran ebookmaker it crashed in the same way.

> Traceback (most recent call last):
>   File "C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\Scripts\ebookmaker", line 16, in <module>
>     from ebookmaker import EbookMaker
>   File "C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\lib\site-packages\ebookmaker\EbookMaker.py", line 33, in <modu
> le>
>     from libgutenberg import Cover
>   File "C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\lib\site-packages\libgutenberg\Cover.py", line 35, in <module>
> 
>     import cairocffi as cairo
>   File "C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\lib\site-packages\cairocffi\__init__.py", line 41, in <module>
> 
>     cairo = dlopen(ffi, 'cairo', 'cairo-2')
>   File "C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\lib\site-packages\cairocffi\__init__.py", line 38, in dlopen
>     raise OSError("dlopen() failed to load a library: %s" % ' / '.join(names))
> OSError: dlopen() failed to load a library: cairo / cairo-2

When I rename the gtk folder back, ebookmaker runs correctly again.

**Comment by Houdinic** [comment]
I updated the file and comment out the lines that generated the errors. Can you try it gain? @windymilla

File: ebookmaker/EbookMaker.py [hunk 3]
```diff
@@ -30,7 +30,7 @@
 from libgutenberg.Logger import debug, warning, error, exception
 from libgutenberg import Logger, DublinCore
 from libgutenberg import MediaTypes as mt
-from libgutenberg import Cover
+#from libgutenberg import Cover
```
Comment (eshellman) [inline ebookmaker/EbookMaker.py:33]: can't do hasattr(Cover, 'cairo') without importing Cover!

Comment (Houdinic) [reply ebookmaker/EbookMaker.py:33]: But if we import Cover, Cover would import cairo which causes the errors. We cannot modify files in libgutenberg for this problem right?

**Comment by windymilla** [comment]
> I updated the file and comment out the lines that generated the errors. Can you try it gain? @windymilla

I tried it and that worked for me even with the gtk folder renamed so Cairo isn't available. The cover provided by the PPer was correctly included.

**Comment by eshellman** [comment]
current code will fail if there's not a provided cover.

based on feedback from @windymilla  (Thanks again!) we should be able to import Cover successfully and run ebookmaker without cairo if we use libgutenberg==0.6.3. (new version from a few minutes ago)

so ...
```
pipenv uninstall ebookmaker
pipenv install 'libgutenberg==0.6.3'
pipenv install -e git+https://github.com/gutenbergtools/ebookmaker.git@winpath#egg=ebookmaker
```

**Comment by windymilla** [comment]
Hi

I uninstalled ebookmaker, installed libgutenberg==0.6.3, then reinstalled ebookmaker.
Although I got what looked like serious errors when installing ebookmaker (see attached out.log), it does seem to add a cover image if one is given, and creates a book without a cover image if one is not there.

However, see next message for another issue relating to working folder (sorry!)
Nigel
[out.log](https://github.com/gutenbergtools/ebookmaker/files/4497049/out.log)

**Comment by windymilla** [comment]
In order to run ebookmaker I have to cd to the folder where I installed it (i.e. where I typed "pipenv install..."). In my case, this is C:\DP\ebookmaker which has a corresponding virtualenv folder "C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi". 

If I try to run ebookmaker from the folder where the HTML file is that I want to convert, then "pipenv run" tries to make a virtual environment for that folder, and can't find ebookmaker.

So, I'm running a command like the following from C:\DP\ebookmaker:
`pipenv run python C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\Scripts\ebookmaker -v --make=epub.images --make=kindle.images --output-dir=C:\DP\myproject --title=mybook C:\DP\myproject\mybook.html`

The problem is that ebookmaker looks for an "images" folder in its working directory, (C:\DP\ebookmaker) when of course in the above example it is C:\DP\myproject\images. It therefore doesn't find the cover or other image files. I didn't notice previously as I was doing the initial tests from the C:\DP\ebookmaker folder, and then when I tested it from a real project folder location, the project was actually the one I had been using for testing, so unknown to me it was getting the images and cover from the ebookmaker folder and everything seemed fine.

I think ebookmaker should extract the directory name of the file it is asked to process, then either make that its current working directory, or add it to the start of the images folder (and possibly other folders/files?)

Sorry for not spotting this sooner - hopefully it's not too hard to get round.
Thanks again @eshellman and @Houdinic 
Nigel

**Comment by Houdinic** [comment]
> In order to run ebookmaker I have to cd to the folder where I installed it (i.e. where I typed "pipenv install..."). In my case, this is C:\DP\ebookmaker which has a corresponding virtualenv folder "C:\Users\nblower.virtualenvs\ebookmaker-cgaQuYhi".
> 
> If I try to run ebookmaker from the folder where the HTML file is that I want to convert, then "pipenv run" tries to make a virtual environment for that folder, and can't find ebookmaker.
> 
> So, I'm running a command like the following from C:\DP\ebookmaker:
> `pipenv run python C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\Scripts\ebookmaker -v --make=epub.images --make=kindle.images --output-dir=C:\DP\myproject --title=mybook C:\DP\myproject\mybook.html`
> 
> The problem is that ebookmaker looks for an "images" folder in its working directory, (C:\DP\ebookmaker) when of course in the above example it is C:\DP\myproject\images. It therefore doesn't find the cover or other image files. I didn't notice previously as I was doing the initial tests from the C:\DP\ebookmaker folder, and then when I tested it from a real project folder location, the project was actually the one I had been using for testing, so unknown to me it was getting the images and cover from the ebookmaker folder and everything seemed fine.
> 
> I think ebookmaker should extract the directory name of the file it is asked to process, then either make that its current working directory, or add it to the start of the images folder (and possibly other folders/files?)
> 
> Sorry for not spotting this sooner - hopefully it's not too hard to get round.
> Thanks again @eshellman and @Houdinic
> Nigel

Thanks for your feedback! We will try to fix this problem.

**Comment by eshellman** [comment]
1. pipenv sets the environment based on the directory its run from.
2. yes, another windows path issue. we just need to apply the fix we did for outputdir to other places - no good excuse for continuing to do it wrong!

**Comment by windymilla** [comment]
Hi
Not sure if it was ready for me to test, but I did anyway - copied command and output from command window into attached log file.
If it's not ready for me, just ignore this, but I've uploaded it in case it's useful.
Nigel
[output.txt](https://github.com/gutenbergtools/ebookmaker/files/4525117/output.txt)

**Comment by eshellman** [comment]
version 0.8.10 is uploaded to PyPI. Will merge tomorrow if nothing else comes up.

0.8.10 April 30, 2020
- fixed numerous file path nits on Windows
    - kindlegen
    - figsdir
    - cover
    - pdf
- PEP8 delinted:
    - EbookMaker.py
    - ParserFactory.py
    - Spider.py
    - parsers/__init__.py
    - writers/EpubWriter.py

0.8.9 April 13, 2020
- fixed issues preventing successful deployments on Windows
- added logging for books deeper than max_depth
- improved documentation for tidy and cairo prerequisites
- try to catch  and report exceptions when tidy and cairo aren't installed
- added install-on-windows notes

**Comment by windymilla** [comment]
Version 0.8.10 works for me:
I can run it from a folder other than the folder linked to the virtual
environment.
I have a simple Windows .bat file that cd's to the virtualenv linked folder
and runs ebookmaker given foldername and filename arguments. This bat file
runs from within Guiguts Custom menu and produces the epub and Kindle
versions in the same folder as the HTML file you are editing - exactly what
GG users would like. It would also be easy for users of other Windows tools
to use a similar .bat file to run ebookmaker on an HTML file.

Minor note:Things go wrong if there is a space in the path to the file,
even with quotes round it, e.g.

C:\DP\ebookmaker> pipenv run python
C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\Scripts\ebookmaker -v
--make=epub.images --make=kindle.images
--output-dir="C:\DP\PPV\QGC
AWeekinWallStreet\projectID5cb349d78a2ec_second\"
--title=wallstreet-utf8 "C:\DP\PPV\QGC
AWeekinWallStreet\projectID5cb349d78a2ec_second\wallstreet-utf8.html"
WARNING:root:No gnu dbm support found. Rhyming dictionary not used.
ERROR:root:Missing file:
C:\DP\ebookmaker\AWeekinWallStreet\projectID5cb349d78a2ec_second\wallstreet-utf8.html
etc.....fatal errors continue from here

The space was after QGC in the foldername. Looks like it interprets the bit
after the space as a relative pathname from the folder that ebookmaker is
installed & run in.

This is minor. If fixing it is not easy, it could just be documented as a
limitation.

There are also lots of errors like the following whatever the folder name,
but they don't seem to be fatal or detrimental.
ERROR:root:Omitted file
file:///C:/DP/PPV/QGCAWeekinWallStreet/projectID5cb349d78a2ec_second/wallstreet-utf8.html
due to depth > max_depth

Many thanks for all your hard work on this!
Nigel

**Comment by eshellman** [comment]
Nigel - I'm shocked that it works with spaces in the name on my mac, but it does. It's a strange way to break!

**Comment by eshellman** [comment]
@windymilla we are unable to reproduce the problem with spaces in the directory names. is the bat file involved?

**Comment by windymilla** [comment]
@eshellman - I've checked using the command in a command prompt window rather than via a .bat file with the same results. 

If there is a space after QGC within the path, then it errors as described above. 
If I then remove the space (from the output dir argument and the final file argument) and also rename the folder to remove the space, then it works fine.

I don't know if the following gives a hint as to the problem: if I put quotes round the path, even without the space, then I get an error which implies it is not recognising the final argument as a filename. Could it be that the quotes have to be somehow handled within the ebookmaker? I'm not sure what argument it receives from the Windows command line - are the quotes stripped or not?

> C:\DP\ebookmaker>pipenv run python C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\Scripts\ebookmaker -v --make=epub.i
> mages --make=kindle.images --output-dir="C:\DP\PPV\QGCAWeekinWallStreet\projectID5cb349d78a2ec_second\" --title=wallstre
> etutf8 "C:\DP\PPV\QGCAWeekinWallStreet\projectID5cb349d78a2ec_second\wallstreetutf8.html"
> 
> usage: EbookMaker [-h] [--verbose] [--config CONFIG_FILE] [--version]
>                   [--make {pdf.noimages,picsdir.noimages,epub.noimages,all,txt.utf-8,kindle,pdf,pdf.images,html,html.noi
> mages,kindle.noimages,html.images,epub.images,picsdir.images,txt.us-ascii,txt.iso-8859-1,kindle.images,epub,rst,test,rst
> .gen,txt}]
>                   [--max-depth LEVELS] [--strip_links] [--include GLOB]
>                   [--exclude GLOB] [--include-mediatype GLOB/GLOB]
>                   [--exclude-mediatype GLOB/GLOB]
>                   [--input-mediatype MEDIATYPE] [--mediatype-from-extension]
>                   [--rewrite from>to] [--title TITLE] [--author AUTHOR]
>                   [--ebook EBOOK] [--output-dir OUTPUT_DIR]
>                   [--output-file OUTPUT_FILE] [--validate]
>                   [--section TAG.CLASS] [--packager {ww,gzip}]
>                   [--cover COVERPAGE_URL] [--generate_cover] [--jobs]
>                   [--extension-package PYTHON_PACKAGE]
>                   url
>
> EbookMaker: error: the following arguments are required: url

**Comment by eshellman** [comment]
@windymilla 0.8.10 broke some things on linux, so we're pushing out another release. 

Instead of ```
--output-dir="C:\DP\PPV\QGC
AWeekinWallStreet\projectID5cb349d78a2ec_second\"
``` could you try these variants:```
--output-dir="C:\DP\PPV\'QGC
AWeekinWallStreet'\projectID5cb349d78a2ec_second\"
``` and ```
--output-dir="C:\DP\PPV\QGC%20AWeekinWallStreet\projectID5cb349d78a2ec_second\"
```

Thanks!

**Comment by windymilla** [comment]
I assumed you wanted me to use those variants in both places the space appears in the command - if not, I can try again. Here are the results:

With single quotes round the part with a space in it (same issue):
C:\DP\ebookmaker>pipenv run python C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\Scripts\ebookmaker -v --make=epub.images --make=kindle.images --output-dir="C:\DP\PPV\'QGC AWeekinWallStreet'\projectID5cb349d78a2ec_second\" --title=wallstreet-utf8 "C:\DP\PPV\'QGC AWeekinWallStreet'\projectID5cb349d78a2ec_second\wallstreet-utf8.html"
WARNING:root:No gnu dbm support found. Rhyming dictionary not used.
ERROR:root:Missing file: C:\DP\ebookmaker\AWeekinWallStreet'\projectID5cb349d78a2ec_second\wallstreet-utf8.html
INFO:root:Continuing parse despite missing file
etc...

Using %20 instead of space (Usage message)
C:\DP\ebookmaker>pipenv run python C:\Users\nblower\.virtualenvs\ebookmaker-cgaQuYhi\Scripts\ebookmaker -v --make=epub.images --make=kindle.images --output-dir="C:\DP\PPV\QGC%20AWeekinWallStreet\projectID5cb349d78a2ec_second\" --title=wallstreet-utf8 "C:\DP\PPV\QGC%20AWeekinWallStreet\projectID5cb349d78a2ec_second\wallstreet-utf8.html"
usage: EbookMaker [-h] [--verbose] [--config CONFIG_FILE] [--version]
etc...

From something I saw online, I also tried preceding the spaces with a caret, i.e. QGC^ AWeekinWallStreet, but that gave the same results as variant 1 above.

Unsurprisingly perhaps, using the short (DOS) pathname works (no quotes), i.e.
--output-dir=C:\DP\PPV\QGCAWE~1\projectID5cb349d78a2ec_second\

Sorry, we still haven't got to the bottom of it. I'm happy to edit "print or debug" commands into the version of ebookmaker I have installed here, if that would help find out how ebookmaker is interpreting the command line under Windows.

Thanks
Nigel

**Comment by Houdinic** [comment]
> I assumed you wanted me to use those variants in both places the space appears in the command - if not, I can try again. Here are the results:
> 
> With single quotes round the part with a space in it (same issue):
> C:\DP\ebookmaker>pipenv run python C:\Users\nblower.virtualenvs\ebookmaker-cgaQuYhi\Scripts\ebookmaker -v --make=epub.images --make=kindle.images --output-dir="C:\DP\PPV'QGC AWeekinWallStreet'\projectID5cb349d78a2ec_second" --title=wallstreet-utf8 "C:\DP\PPV'QGC AWeekinWallStreet'\projectID5cb349d78a2ec_second\wallstreet-utf8.html"
> WARNING:root:No gnu dbm support found. Rhyming dictionary not used.
> ERROR:root:Missing file: C:\DP\ebookmaker\AWeekinWallStreet'\projectID5cb349d78a2ec_second\wallstreet-utf8.html
> INFO:root:Continuing parse despite missing file
> etc...
> 
> Using %20 instead of space (Usage message)
> C:\DP\ebookmaker>pipenv run python C:\Users\nblower.virtualenvs\ebookmaker-cgaQuYhi\Scripts\ebookmaker -v --make=epub.images --make=kindle.images --output-dir="C:\DP\PPV\QGC%20AWeekinWallStreet\projectID5cb349d78a2ec_second" --title=wallstreet-utf8 "C:\DP\PPV\QGC%20AWeekinWallStreet\projectID5cb349d78a2ec_second\wallstreet-utf8.html"
> usage: EbookMaker [-h] [--verbose] [--config CONFIG_FILE] [--version]
> etc...
> 
> From something I saw online, I also tried preceding the spaces with a caret, i.e. QGC^ AWeekinWallStreet, but that gave the same results as variant 1 above.
> 
> Unsurprisingly perhaps, using the short (DOS) pathname works (no quotes), i.e.
> --output-dir=C:\DP\PPV\QGCAWE~1\projectID5cb349d78a2ec_second\
> 
> Sorry, we still haven't got to the bottom of it. I'm happy to edit "print or debug" commands into the version of ebookmaker I have installed here, if that would help find out how ebookmaker is interpreting the command line under Windows.
> 
> Thanks
> Nigel

We have posted new version of ebookmaker. Can you have a try on that? It's version 0.8.12

**Comment by windymilla** [comment]
Sorry guys - 0.8.12 is still doing the same thing. Either a "Usage" message
when trying %20, or it takes the final argument as it sees it:
  AWeekinWallStreet'\projectID5cb349d78a2ec_second\wallstreet-utf8.html
as being a relative path, so complains that it can't find
 C:\DP\ebookmaker\AWeekinWallStreet'\projectID5cb349d78a2ec_second\wallstreet-utf8.html
I tried with and without double quotes on the whole argument, with and
without single quotes around the bit with a space, %20 and caret methods,
but no luck.
Nigel

**Pull request merged**